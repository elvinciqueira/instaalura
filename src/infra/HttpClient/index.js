export default async function HttpClient(
  url,
  {headers, body, ...options},
  fetchModule = fetch,
) {
  const serverResponse = await fetchModule(url, {
    headers: {
      ...headers,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
    ...options,
  })

  if (serverResponse.ok) {
    return serverResponse.json()
  }

  throw new Error('Não foi possível cadastrar o usuário agora :(')
}
