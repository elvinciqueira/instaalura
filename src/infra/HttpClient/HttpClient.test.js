import HttpClient from './index'

const fetchModule = jest.fn()

describe('HttpClient()', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  describe('when HttpClient gets called', () => {
    describe('and succeed', () => {
      it('returns a server response', async () => {
        const url = 'some-url/pi'
        const body = {
          username: 'some-username',
          password: 'some-password',
        }
        const requestedParameters = {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(body),
        }

        fetchModule.mockImplementationOnce(() =>
          Promise.resolve({
            ok: true,
            json: () => ({
              data: {
                username: 'some-username',
              },
            }),
          }),
        )

        const serverResponse = await HttpClient(
          url,
          {
            method: 'POST',
            body: {
              username: 'some-username',
              password: 'some-password',
            },
          },
          fetchModule,
        )

        expect(fetchModule).toHaveBeenCalledTimes(1)
        expect(fetchModule).toHaveBeenCalledWith(url, requestedParameters)
        expect(serverResponse).toMatchSnapshot()
      })
    })

    describe('and fails', () => {
      it('throws an error', async () => {
        fetchModule.mockImplementationOnce(() =>
          Promise.reject('Não foi possível cadastrar o usuário agora :('),
        )

        expect.assertions(2)

        try {
          await HttpClient('some-url/api', {}, fetchModule)
        } catch (error) {
          expect(error).toMatch('Não foi possível cadastrar o usuário agora :(')
        }

        expect(fetchModule).toHaveBeenCalledTimes(1)
      })
    })
  })
})
