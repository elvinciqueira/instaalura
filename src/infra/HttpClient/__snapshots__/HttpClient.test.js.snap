// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpClient() when HttpClient gets called and succeed returns a server response 1`] = `
Object {
  "data": Object {
    "username": "some-username",
  },
}
`;
